File: app/src/main/kotlin/metamark/parser/DefinitionTerm.kt
package metamark.parser

import metamark.parser.antlr.MetamarkLexer
import metamark.parser.antlr.MetamarkParser
import metamark.parser.antlr.MetamarkParser.Expression_stringLiteralContext

sealed class DefinitionTerm {
    companion object {
        fun build(ctx: MetamarkParser.DefinitionContext): DefinitionTerm {
            val kindToken = ctx.kind

            return when (kindToken.type) {
                MetamarkLexer.SymbolKeyword -> SymbolDefinitionTerm.build(ctx)
                MetamarkLexer.RuleKeyword -> RuleDefinitionTerm.build(ctx)
                else -> throw IllegalArgumentException(kindToken.toString())
            }
        }
    }

    abstract val name: String
}

data class SymbolDefinitionTerm(
    override val name: String,
    val pattern: StringLiteralTerm,
) : DefinitionTerm() {
    companion object {
        fun build(ctx: MetamarkParser.DefinitionContext): DefinitionTerm {
            val body = ctx.body

            return if (body is Expression_stringLiteralContext) {
                SymbolDefinitionTerm(
                    name = ctx.name.text, pattern = StringLiteralTerm.build(body.stringLiteral())
                )
            } else {
                throw IllegalArgumentException(body::class.java.toString())
            }
        }
    }
}

data class RuleDefinitionTerm(
    override val name: String,
    val body: ExpressionTerm,
) : DefinitionTerm() {
    companion object {
        fun build(ctx: MetamarkParser.DefinitionContext): DefinitionTerm {
            return RuleDefinitionTerm(
                name = ctx.name.text,
                body = ExpressionTerm.build(ctx.body),
            )
        }
    }
}

File: src/main/DefinitionTerm.ts
import {DefinitionContext, MetamarkParser, StringLiteralContext} from "../generated_parser/MetamarkParser";
import {MetamarkLexer} from "../generated_parser/MetamarkLexer";
import {ExpressionTerm} from "./ExpressionTerm";
import {StringLiteralTerm} from "./StringLiteralTerm";

export abstract class DefinitionTerm {
    static build(ctx: DefinitionContext): DefinitionTerm {
        if (ctx._kind.type === MetamarkLexer.SymbolKeyword) {
            return new SymbolDefinitionTerm(
                ctx._name.text,
                StringLiteralTerm.build(ctx._body as StringLiteralContext),
            );
        } else if (ctx._kind.type === MetamarkLexer.RuleKeyword) {
            return new RuleDefinitionTerm(
                ctx._name.text,
                ExpressionTerm.build(ctx._body),
            );
        } else {
            throw new Error(`Unsupported definition kind token: ${ctx._kind}`);
        }
    }

    abstract readonly name: string;
}

export class SymbolDefinitionTerm extends DefinitionTerm {
    override readonly name: string;
    readonly pattern: StringLiteralTerm;

    constructor(name: string, pattern: StringLiteralTerm) {
        super();
        this.name = name;
        this.pattern = pattern;
    }
}

export class RuleDefinitionTerm extends DefinitionTerm {
    override readonly name: string;
    readonly body: ExpressionTerm;

    constructor(name: string, body: ExpressionTerm) {
        super();
        this.name = name;
        this.body = body;
    }
}
